flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -621,6 +692,237 @@ private static void checkInterrupted() throws InterruptedException {
+			if (backupFileSystem.exists(backupPath)) {
+				LOG.warn("Deleting an existing local checkpoint directory " +
+					backupPath + ".");
+
+				backupFileSystem.delete(backupPath, true);
+			}
+
+			// create hard links of living files in the checkpoint path
+			Checkpoint checkpoint = Checkpoint.create(stateBackend.db);
+			checkpoint.createCheckpoint(backupPath.getPath());
+		}
+
+		KeyedStateHandle materializeSnapshot() throws Exception {
+			// write meta data
+			metaStateHandle = materializeMetaData();
+
+			// write state data
+			Preconditions.checkState(backupFileSystem.exists(backupPath));
+
+			FileStatus[] fileStatuses = backupFileSystem.listStatus(backupPath);
+			if (fileStatuses != null ) {
+				for (FileStatus fileStatus : fileStatuses) {
+					Path filePath = fileStatus.getPath();
+					String fileName = filePath.getName();
+
+					if (fileName.endsWith(SST_FILE_SUFFIX)) {
+						StreamStateHandle fileHandle =
+							baseSstFiles == null ? null : baseSstFiles.get(fileName);
+
+						if (fileHandle == null) {
+							newSstFileNames.add(fileName);
+							fileHandle = materializeStateData(filePath);
+						}
+
+						sstFiles.put(fileName, fileHandle);
+					} else {
+						StreamStateHandle fileHandle = materializeStateData(filePath);
+						miscFiles.put(fileName, fileHandle);
+					}
+				}
+			}
+
+			stateBackend.materializedSstFiles.put(checkpointId, sstFiles);
+
+			return new RocksDBKeyedStateHandle(stateBackend.jobId,
+				stateBackend.operatorIdentifier, stateBackend.keyGroupRange,
+				newSstFileNames, sstFiles, miscFiles, metaStateHandle);
+		}
+
+		void releaseResources(boolean canceled) {
+
+			if (inputStream != null) {
+				stateBackend.cancelStreamRegistry.unregisterClosable(inputStream);
+				try {
+					inputStream.close();
+				} catch (Exception e) {
+					LOG.warn("Could not properly close the input stream.", e);
+				}
+				inputStream = null;
+			}
+
+			if (outputStream != null) {
+				stateBackend.cancelStreamRegistry.unregisterClosable(outputStream);
+				try {
+					outputStream.close();
+				} catch (Exception e) {
+					LOG.warn("Could not properly close the output stream.", e);
+				}
+				outputStream = null;
+			}
+
+			if (backupPath != null) {
+				try {
+					if (backupFileSystem.exists(backupPath)) {
+						backupFileSystem.delete(backupPath, true);
+					}
+				} catch (Exception e) {
+					LOG.warn("Could not properly delete the checkpoint directory.", e);
+				}
+			}
+
+			if (canceled) {
+				List<StateObject> statesToDiscard = new ArrayList<>();
+
+				if (metaStateHandle != null) {
+					statesToDiscard.add(metaStateHandle);
+				}
+
+				statesToDiscard.addAll(miscFiles.values());
+
+				for (String newSstFileName : newSstFileNames) {
+					StreamStateHandle fileHandle = sstFiles.get(newSstFileName);
+					if (fileHandle != null) {
+						statesToDiscard.add(fileHandle);
+					}
+				}
+
+				try {
+					StateUtil.bestEffortDiscardAllStateObjects(statesToDiscard);
```

Comment:
You are absolutely right. Global cleanup hooks are urgently needed here to clean unused states here. The hook at job manager instead of the shared state registry is supposed to do the work because unused private states should be cleaned as well.   At the startup, the hook will know the checkpoint from which we are restoring and only retain the data in restored completed checkpoints.

The local checkpointing directories will be deleted once the checkpoint completed at the TM. Since local checkpoint directories are all under the directory for the backend which is deleted when the backend is disposed, they can also be deleted if the backend is correctly closed. 

But in the cases when the TM fails during the closing of the backend, the local checkpoint directories will be left on the file system. The problem does not matter in Yarn clusters but may be very severe in standalone clusters. What do you think of the problem?